trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/expressiontranslation/BitabsTranslation.java

```diff
@@ -30,384 +30,360 @@
-	private final FunctionDeclarations mFunctionDeclarations;
 	private final TypeSizes mTypeSizes;
-	private final ITypeHandler mTypeHandler;
 
-	public BitabsTranslation(final TypeSizes typeSizes, final ITypeHandler typeHandler,
-			final FunctionDeclarations functionDeclarations) {
+	public BitabsTranslation(final TypeSizes typeSizes) {
 		mTypeSizes = typeSizes;
-		mTypeHandler = typeHandler;
-		mFunctionDeclarations = functionDeclarations;
 	}
 
-	public Expression abstractAnd(final ILocation loc, final Expression left, final CPrimitive typeLeft,
-			final Expression right, final CPrimitive typeRight) {
+	/**
+	 * Overapproximates the bitwise {@code and}. Uses the following rules to increase the precision:
+	 * <li>0 & a = a & 0 = 0
+	 * <li>a & a = a
+	 * <li>If a >= 0 or b < 0, then a & b <= a
+	 * <li>If a < 0 or b >= 0, then a & b <= b
+	 * <li>If a >= b or b >= 0, then a & b >= 0
+	 * <li>If a < 0 or b < 0, then a & b > a + b
+	 */
+	public ExpressionResult abstractAnd(final ILocation loc, final Expression left, final Expression right,
+			final CPrimitive type, final AuxVarInfoBuilder auxVarInfoBuilder) {
 		// 0 & a = a & 0 = 0
-		if (left instanceof IntegerLiteral && ((IntegerLiteral) left).getValue().equals("0")) {
-			return left;
+		if (isZero(left)) {
+			return new ExpressionResult(new RValue(left, type));
 		}
-		if (right instanceof IntegerLiteral && ((IntegerLiteral) right).getValue().equals("0")) {
-			return right;
+		if (isZero(right)) {
+			return new ExpressionResult(new RValue(right, type));
 		}
-		// a & a = a
-		if (left instanceof IntegerLiteral && right instanceof IntegerLiteral
-				&& ((IntegerLiteral) left).getValue().equals(((IntegerLiteral) right).getValue())) {
-			return left;
+		if (left instanceof IntegerLiteral && right instanceof IntegerLiteral) {
+			return handleConstants((IntegerLiteral) left, (IntegerLiteral) right, BigInteger::and, loc, type);
 		}
 		final Expression zero = new IntegerLiteral(loc, BoogieType.TYPE_INT, "0");
 
-		final Expression leftEqualsZero = ExpressionFactory.newBinaryExpression(loc, Operator.COMPEQ, left, zero);
-		final Expression rightEqualsZero = ExpressionFactory.newBinaryExpression(loc, Operator.COMPEQ, right, zero);
+		final AuxVarInfo auxvarinfo = auxVarInfoBuilder.constructAuxVarInfo(loc, type, SFO.AUXVAR.NONDET);
+
+		final Expression auxvar = applyWraparoundIfNecessary(loc, auxvarinfo.getExp(), type);
+		final Expression leftWrapped = applyWraparoundIfNecessary(loc, left, type);
+		final Expression rightWrapped = applyWraparoundIfNecessary(loc, right, type);
+
+		final Expression smallerLeft =
+				ExpressionFactory.newBinaryExpression(loc, Operator.COMPLEQ, auxvar, leftWrapped);
+		final Expression smallerRight =
+				ExpressionFactory.newBinaryExpression(loc, Operator.COMPLEQ, auxvar, rightWrapped);
+
+		final List<Expression> assumptions;
+		if (mTypeSizes.isUnsigned(type)) {
+			assumptions = List.of(smallerLeft, smallerRight);
+		} else {
```

Comment:
This kind of code (in this class generally) is the reason why I think a simpler DSL to create Boogie expressions would make working on the C translation so much easier, and the code so much more understandable.

(But this is not something we can fix in this PR, just a note.)